Update component dependencies docs for v4 (#201)
spinframework/spin-docs: content/v4/writing-apps.md

Few of us write applications without relying on libraries. Traditionally, those libraries have had to come from the language ecosystem - e.g. `npm` for JavaScript, `pip` for Python, etc. - and you can still work this way in Spin. However, the WebAssembly Component Model means that you can also depend on other WebAssembly components. The process of combining your application component with the Wasm components it depends on is called _composition_, and Spin supports this natively.

> > Spin composes each of your component's imports independently against a single dependency: the dependency component is not further composed with any other components. Spin also wires up the dependency's own capability imports selectively, based on the permissions you explicitly grant it (see [Dependency Permissions](#dependency-permissions)) — any capability you do not grant is satisfied by a stub that returns an error at runtime, rather than being connected to the host. If you need to construct a more complex composition, you must use a dedicated tool such as [`wac`](https://github.com/bytecodealliance/wac) as part of your build. See the [Component Model book](https://component-model.bytecodealliance.org/creating-and-consuming/composing.html) for details and examples.

Reviewer: This feels like like two paragraphs, one about "use wac if plug doesn't suffice" (which was the intent of the previous version), and one about selective permissions.  I agree the existing text could be tightened to focus more clearly on the, uh, focus, but I don't think we need to talk about selective permissions (and the deny adapter implementation) as part of it; and I feel that doing so makes the point of the paragraph even harder to follow.

If we do want to talk about the deny adapter, I suggest we make that a "how it works" note in the dependency permissions section.  But again, not sure we need to do into that.

Author: Yeah i agree. I was just wanting to communicate away from "plug" and socialize the idea of selective imports but there are probably better ways than re-formulating this section.

### Dependency Permissions

> By default, dependencies do not have access to Spin resources that require permission to be given in the manifest - network hosts, key-value stores, SQLite databases, variables, etc. Any capability imports a dependency makes that you have not granted are wired to stub implementations that return errors at runtime, so an over-reaching dependency will fail loudly rather than silently bypass the parent's configuration.

Reviewer: ```suggestion
By default, dependencies do not have access to Spin resources that require permission to be given in the manifest - network hosts, key-value stores, SQLite databases, variables, etc. If a dependency tries to access such a capability, it will fail with an error.
```

You control this per dependency with the `inherit_configuration` field, which accepts three forms:

> - A list of capability names — the dependency is granted only those capabilities.

Reviewer: This seems like it kind of restates the next three-ish paragraphs?

| `variables` | The dependency can read all Spin configuration variables listed in the component manifest. The dependency may use the [variables API](./variables.md). |

`inherit_configuration` applies uniformly to every dependency source — registry packages, local paths, HTTP URLs, and in-app component references. For example, the same permission form works for a local dependency:

```toml
[component.my-app.dependencies]
"my:dependency" = { path = "../deps/my-dependency.wasm", inherit_configuration = ["allowed_outbound_hosts"] }
```

#### Granting all or no capabilities

To grant the dependency the same permissions as the main component, set `inherit_configuration` to `true`:

```toml
[component.my-app.dependencies]
"highly:trusted/dependency" = { version = "1.0.0", inherit_configuration = true }
```

To make the isolation explicit (for example, in a manifest where other dependencies do inherit), set `inherit_configuration` to `false`. This is identical to omitting the field:

```toml
[component.my-app.dependencies]
"untrusted:dependency" = { version = "1.0.0", inherit_configuration = false }
```

> #### Granting all dependencies the same permissions

Reviewer: ```suggestion
#### Granting all dependencies full permissions
```

it's a bit more than "all deps the same" - not a blocker though

| `sqlite_databases` | The dependency can access all SQLite databases listed in the component manifest. The dependency may use the [SQLite API](./sqlite-api-guide.md). |
| `variables` | The dependency can read all Spin configuration variables listed in the component manifest. The dependency may use the [variables API](./variables.md). |

> The dependency does not receive any permissions other than those you list. Any other capability imports the dependency makes will return errors at runtime.

Reviewer: This is mentioned above and feels kind of implicit.  The reader can probably guess what happens if a dependency tries to use an unpermitted thing!

`inherit_configuration` applies uniformly to every dependency source — registry packages, local paths, HTTP URLs, and in-app component references. For example, the same permission form works for a local dependency:

```toml
[component.my-app.dependencies]
"my:dependency" = { path = "../deps/my-dependency.wasm", inherit_configuration = ["allowed_outbound_hosts"] }
```

> > The shorthand version-string form (`"my:dependency" = "1.0.0"`) cannot carry `inherit_configuration`. Use the expanded table form whenever you need to grant permissions.

Reviewer: This seems implicit in the syntax.  If I had a version string, how would I set `inherit_configuration`?  Well I'd need to add another field, so it would need to be a table...

(Also the whole "address it by version string alone" depends - unless I have misunderstood - on a whole bunch of implicit default registry stuff that seems pretty rare at the moment, so I don't particularly want to put focus on this syntax anyway...!)

"my:dependency" = "1.0.0"  # has all capabilities of the main component
```

> > `dependencies_inherit_configuration` and per-dependency `inherit_configuration` are **mutually exclusive** on the same component. The following manifest is an error:

Reviewer: We don't usually dig into "you can't do both X and Y" cases in this much detail unless we hear that users are doing it.  In this case it seems unlikely that someone is going to try to use both, and if they do, the error message is clear.  I think in this case it's fine to trust them to figure it out.

Author: Fair. I added this because as a dev one of the first things I would try to do is attempt to use both out of morbid curiosity. However I recognize my brain is weird. 🤣
